server/test/unit/server/content/test_storage.py

```diff
@@ -62,45 +62,53 @@ def test_exit(self):
 
 class TestFileStorage(TestCase):
 
-    @patch('pulp.server.content.storage.shutil')
     @patch('pulp.server.content.storage.config')
-    @patch('os.path.isdir', Mock(return_value=True))
-    def test_put_dir(self, config, shutil):
-        path_in = '/tmp/test/'
+    def test_get_path(self, config):
         storage_dir = '/tmp/storage'
-        unit = Mock(id='0123456789', _content_type_id='ABC')
         config.get = lambda s, p: {'server': {'storage_dir': storage_dir}}[s][p]
+        unit = Mock(id='0123456789', type_id='ABC')
+
+        # test
+        path = FileStorage.get_path(unit)
+
+        # validation
+        self.assertEqual(
+            path,
+            os.path.join(os.path.join(storage_dir, 'content', 'units'),
+                         unit.type_id,
+                         unit.id[0:4],
+                         unit.id))
+
+    @patch('pulp.server.content.storage.shutil')
+    @patch('pulp.server.content.storage.mkdir')
+    def test_put_file(self, _mkdir, shutil):
+        path_in = '/tmp/test'
+        unit = Mock(id='123', storage_path='/tmp/storage')
         storage = FileStorage()
 
         # test
         storage.put(unit, path_in)
 
         # validation
-        destination = os.path.join(
-            os.path.join(storage_dir, 'content', 'units', unit._content_type_id),
-            unit.id[0:4], unit.id)
-        shutil.copytree.assert_called_once_with(path_in, destination)
-        self.assertEqual(unit._storage_path, destination)
+        _mkdir.assert_called_once_with(os.path.dirname(unit.storage_path))
+        shutil.copy2.assert_called_once_with(path_in, unit.storage_path)
 
     @patch('pulp.server.content.storage.shutil')
-    @patch('pulp.server.content.storage.config')
-    @patch('os.path.isdir', Mock(return_value=False))
-    def test_put_file(self, config, shutil):
+    @patch('pulp.server.content.storage.mkdir')
+    def test_put_file_with_location(self, _mkdir, shutil):
         path_in = '/tmp/test'
-        storage_dir = '/tmp/storage'
-        unit = Mock(id='0123456789', _content_type_id='ABC')
-        config.get = lambda s, p: {'server': {'storage_dir': storage_dir}}[s][p]
+        location = '/a/b/'
+        unit = Mock(id='123', _storage_path='/tmp/storage')
         storage = FileStorage()
 
         # test
-        storage.put(unit, path_in)
+        storage.put(unit, path_in, location)
 
         # validation
-        destination = os.path.join(
-            os.path.join(storage_dir, 'content', 'units', unit._content_type_id),
-            unit.id[0:4], unit.id)
-        shutil.copy.assert_called_once_with(path_in, destination)
-        self.assertEqual(unit._storage_path, destination)
+        location.lstrip('/')
```

Comment:
I believe this is a no-op, since ``lstrip`` returns a copy of the string and does not modify the string pointed to by ``location``:
```
In [1]: x = 'hi'

In [2]: x.lstrip('h')
Out[2]: 'i'

In [3]: x
Out[3]: 'hi'
```